### PR TITLE
TOK-421 Introduce optional dids on some methods

### PR DESCRIPTION
**Pull request by KitHat**
Signed-off-by: Nikita Khateev <nikita.khateev@dsr-corporation.com>

### CODE REVIEW

File: src/ledger.rs [hunk 1]
```diff
@@ -368,14 +368,14 @@ impl Ledger {
...
         Ledger::_build_get_ddo_request(command_handle, submitter_did, target_did, cb)
     }
 
-    fn _build_get_ddo_request(command_handle: IndyHandle, submitter_did: &str, target_did: &str, cb: Option<ResponseStringCB>) -> ErrorCode {
-        let submitter_did = c_str!(submitter_did);
+    fn _build_get_ddo_request(command_handle: IndyHandle, submitter_did: Option<&str>, target_did: &str, cb: Option<ResponseStringCB>) -> ErrorCode {
+        let submitter_did = opt_c_str!(submitter_did);
         let target_did = c_str!(target_did);
 
         ErrorCode::from(unsafe { ledger::indy_build_get_ddo_request(command_handle, submitter_did.as_ptr(), target_did.as_ptr(), cb) })
```
Comment (mikelodder7) [inline src/ledger.rs]: This won't work, you need to put.
```
let submitter_did_str = opt_c_str!(submitter_did);
```

then in unsafe put
```
opt_c_ptr!(submitter_did, submitter_did_str)
```

File: src/ledger.rs [hunk 2]
```diff
@@ -520,14 +520,14 @@ impl Ledger {
...
         Ledger::_build_get_nym_request(command_handle, submitter_did, target_did, cb)
     }
 
-    fn _build_get_nym_request(command_handle: IndyHandle, submitter_did: &str, target_did: &str, cb: Option<ResponseStringCB>) -> ErrorCode {
-        let submitter_did = c_str!(submitter_did);
+    fn _build_get_nym_request(command_handle: IndyHandle, submitter_did: Option<&str>, target_did: &str, cb: Option<ResponseStringCB>) -> ErrorCode {
+        let submitter_did = opt_c_str!(submitter_did);
         let target_did = c_str!(target_did);
 
         ErrorCode::from(unsafe { ledger::indy_build_get_nym_request(command_handle, submitter_did.as_ptr(), target_did.as_ptr(), cb) })
```
Comment (mikelodder7) [inline src/ledger.rs]: See previous

File: src/ledger.rs [hunk 3]
```diff
@@ -587,14 +587,14 @@ impl Ledger {
...
         Ledger::_build_get_txn_request(command_handle, submitter_did, ledger_type, seq_no, cb)
     }
 
-    fn _build_get_txn_request(command_handle: IndyHandle, submitter_did: &str, ledger_type: Option<&str>, seq_no: i32, cb: Option<ResponseStringCB>) ->  ErrorCode {
-        let submitter_did = c_str!(submitter_did);
+    fn _build_get_txn_request(command_handle: IndyHandle, submitter_did: Option<&str>, ledger_type: Option<&str>, seq_no: i32, cb: Option<ResponseStringCB>) ->  ErrorCode {
+        let submitter_did = opt_c_str!(submitter_did);
         let ledger_type_str = opt_c_str!(ledger_type);
 
         ErrorCode::from(unsafe { ledger::indy_build_get_txn_request(command_handle, submitter_did.as_ptr(), opt_c_ptr!(ledger_type, ledger_type_str), seq_no, cb) })
```
Comment (mikelodder7) [inline src/ledger.rs]: `submitter_did` should look like `ledger_type`

File: src/ledger.rs [hunk 4]
```diff
@@ -727,14 +727,14 @@ impl Ledger {
...
+    pub fn build_get_attrib_request_async<F: 'static>(submitter_did: Option<&str>, target_did: &str, raw: Option<&str>, hash: Option<&str>, enc: Option<&str>, closure: F) -> ErrorCode where F: FnMut(ErrorCode, String) + Send {
         let (command_handle, cb) = ClosureHandler::convert_cb_ec_string(Box::new(closure));
 
         Ledger::_build_get_attrib_request(command_handle, submitter_did, target_did, raw, hash, enc, cb)
     }
 
-    fn _build_get_attrib_request(command_handle: IndyHandle, submitter_did: &str, target_did: &str, raw: Option<&str>, hash: Option<&str>, enc: Option<&str>, cb: Option<ResponseStringCB>) -> ErrorCode {
-        let submitter_did = c_str!(submitter_did);
+    fn _build_get_attrib_request(command_handle: IndyHandle, submitter_did: Option<&str>, target_did: &str, raw: Option<&str>, hash: Option<&str>, enc: Option<&str>, cb: Option<ResponseStringCB>) -> ErrorCode {
+        let submitter_did = opt_c_str!(submitter_did);
```
Comment (mikelodder7) [inline src/ledger.rs]: See above

File: src/ledger.rs [hunk 5]
```diff
@@ -870,14 +870,14 @@ impl Ledger {
...
+    pub fn build_get_schema_request_async<F: 'static>(submitter_did: Option<&str>, id: &str, closure: F) -> ErrorCode where F: FnMut(ErrorCode, String) + Send {
         let (command_handle, cb) = ClosureHandler::convert_cb_ec_string(Box::new(closure));
 
         Ledger::_build_get_schema_request(command_handle, submitter_did, id, cb)
     }
 
-    fn _build_get_schema_request(command_handle: IndyHandle, submitter_did: &str, id: &str, cb: Option<ResponseStringCB>) -> ErrorCode {
-        let submitter_did = c_str!(submitter_did);
+    fn _build_get_schema_request(command_handle: IndyHandle, submitter_did: Option<&str>, id: &str, cb: Option<ResponseStringCB>) -> ErrorCode {
+        let submitter_did = opt_c_str!(submitter_did);
```
Comment (mikelodder7) [inline src/ledger.rs]: see above

File: src/ledger.rs [hunk 6]
```diff
@@ -1084,14 +1084,14 @@ impl Ledger {
...
+    pub fn build_get_cred_def_request_async<F: 'static>(submitter_did: Option<&str>, id: &str, closure: F) -> ErrorCode where F: FnMut(ErrorCode, String) + Send {
         let (command_handle, cb) = ClosureHandler::convert_cb_ec_string(Box::new(closure));
 
         Ledger::_build_get_cred_def_request(command_handle, submitter_did, id, cb)
     }
 
-    fn _build_get_cred_def_request(command_handle: IndyHandle, submitter_did: &str, id: &str, cb: Option<ResponseStringCB>) -> ErrorCode {
-        let submitter_did = c_str!(submitter_did);
+    fn _build_get_cred_def_request(command_handle: IndyHandle, submitter_did: Option<&str>, id: &str, cb: Option<ResponseStringCB>) -> ErrorCode {
+        let submitter_did = opt_c_str!(submitter_did);
```
Comment (mikelodder7) [inline src/ledger.rs]: see above

File: src/ledger.rs [hunk 7]
```diff
@@ -1733,14 +1733,14 @@ impl Ledger {
...
+    pub fn build_get_revoc_reg_def_request_async<F: 'static>(submitter_did: Option<&str>, id: &str, closure: F) -> ErrorCode where F: FnMut(ErrorCode, String) + Send {
         let (command_handle, cb) = ClosureHandler::convert_cb_ec_string(Box::new(closure));
 
         Ledger::_build_get_revoc_reg_def_request(command_handle, submitter_did, id, cb)
     }
 
-    fn _build_get_revoc_reg_def_request(command_handle: IndyHandle, submitter_did: &str, id: &str, cb: Option<ResponseStringCB>) -> ErrorCode {
-        let submitter_did = c_str!(submitter_did);
+    fn _build_get_revoc_reg_def_request(command_handle: IndyHandle, submitter_did: Option<&str>, id: &str, cb: Option<ResponseStringCB>) -> ErrorCode {
+        let submitter_did = opt_c_str!(submitter_did);
```
Comment (mikelodder7) [inline src/ledger.rs]: see above

**Comment by mikelodder7** [comment]
The same pattern should be applied everywhere

**Comment by KitHat** [comment]
Fixed comments of @mikelodder7, should be fine now

**Pull request merged**